symbolic-base/src/ZkFold/Symbolic/Data/Class.hs

```diff
@@ -4,287 +4,160 @@
-import           Control.DeepSeq             (NFData1)
-import           Data.Bifunctor              (bimap)
-import           Data.Binary                 (Binary)
-import           Data.Function               (flip, ($), (.))
-import           Data.Functor                ((<$>))
-import           Data.Functor.Rep            (Representable)
-import qualified Data.Functor.Rep            as R
-import           Data.Kind                   (Type)
-import           Data.Traversable            (Traversable)
-import           Data.Tuple                  (fst)
-import           Data.Type.Equality          (type (~))
-import           Data.Typeable               (Proxy (..))
-import           GHC.Generics                (U1 (..), (:*:) (..), (:.:) (..))
-import qualified GHC.Generics                as G
-
-import           ZkFold.Algebra.Number       (KnownNat)
-import           ZkFold.Control.HApplicative (hliftA2, hpure)
-import           ZkFold.Data.ByteString      (Binary1)
-import           ZkFold.Data.HFunctor        (hmap)
-import           ZkFold.Data.Orphans         ()
-import           ZkFold.Data.Package         (pack)
-import           ZkFold.Data.Product         (fstP, sndP)
-import           ZkFold.Data.Vector          (Vector)
-import           ZkFold.Symbolic.Class       (Symbolic (WitnessField))
-
-type PayloadFunctor f = (Representable f, Binary (R.Rep f))
-
-type LayoutFunctor f = (Binary1 f, NFData1 f, PayloadFunctor f, Traversable f)
+import           Control.DeepSeq              (NFData, NFData1)
+import           Data.Binary                  (Binary)
+import           Data.Eq                      (Eq)
+import           Data.Function                (id, (.))
+import           Data.Functor                 (fmap)
+import           Data.Functor.Classes         (Eq1, Show1)
+import           Data.Functor.Rep             (Representable)
+import qualified Data.Functor.Rep             as R
+import           Data.Kind                    (Type)
+import           Data.Traversable             (Traversable)
+import           Data.Type.Equality           (type (~))
+import           GHC.Generics                 (V1, (:*:) (..), (:.:) (..))
+import qualified GHC.Generics                 as G
+import           GHC.Show                     (Show)
+
+import           ZkFold.Algebra.Number        (KnownNat)
+import           ZkFold.Control.HApplicative  (hpair)
+import           ZkFold.Data.ByteString       (Binary1)
+import           ZkFold.Data.HFunctor         (hmap)
+import           ZkFold.Data.HFunctor.Classes (HEq, HNFData, HShow)
+import           ZkFold.Data.Orphans          ()
+import           ZkFold.Data.Package          (pack, unpack)
+import           ZkFold.Data.Product          (fstP, sndP)
+import           ZkFold.Data.Vector           (Vector)
+import           ZkFold.Symbolic.Class        (Symbolic (..), embedW)
+
+-- | A newtype wrapper for deriving `Generic1` for symbolic data types.
+newtype Sym f c = Sym { runSym :: c f }
```

Comment:
Looks similar to `Vec` from `ZkFold.Symbolic.Data.Vec`, maybe we can merge them in a single datatype?